joku/cogs/games/stocks.py

```diff
@@ -175,6 +175,97 @@ def _identify_stock(self, channels: typing.Sequence[discord.abc.GuildChannel], n
+
+            ts = datetime.datetime(year=2008, month=7, day=day, hour=hour, minute=minute, second=second)
+
+            em = discord.Embed(title="1st Stock Market of Joku")
+            em.description = "Your asset value is based on best estimates from current market prices."
+            em.set_author(name=user.display_name)
+
+            stocks = await ctx.bot.database.get_user_stocks(user)
+            em.add_field(name="Shares held", value=sum(userstock.amount for userstock in stocks))
+            em.add_field(name="Asset values", value="§{:.2f}".format(sum(userstock.amount * userstock.stock.price for userstock in stocks)))
+
+            em.set_thumbnail(url=user.avatar_url)
+            em.timestamp = ts
+            em.set_footer(text="Thanks capitalism!")
+            em.colour = user.colour
+            await ctx.send(embed=em)
+        else:
+            await ctx.channel.send(
+                  "User **{}** has `§{}` assets worth `§{}`.".format(user, 
+                      sum(userstock.amount for userstock in stocks), 
+                      sum(userstock.amount * userstock.stock.price for userstock in stocks)))
+
+    @assets.command(pass_context=True, aliases=["top", "leaderboard"])
+    async def richest(self, ctx: Context, *, what: str = "value"):
+        """
+        Shows the top 10 stock holders in this guild by number of assets held.
+        """
+        guild = await ctx.bot.database.get_or_create_guild(ctx.guild)
+        if not guild.stocks_enabled:
+            await ctx.send(":x: Stocks are not enabled for this server.")
+            return
+        
+        users = await ctx.bot.database.get_multiple_users(*ctx.message.guild.members)
+        
+        #order users by stock amount or asset worth
+        if what == "amount" or "owned": #order by stock amount
+            sorted_list = list(sorted(users, key=lambda user: sum(userstock.amount for userstock in await ctx.bot.database.get_user_stocks(user))))
+        else:
+            sorted_list = list(sorted(users, key=lambda user: sum(userstock.amount * userstock.stock.price for userstock in await ctx.bot.database.get_user_stocks(user))))
+
+        base = "**Top 10 users (in this server):**\n\n```{}```"
+
+        # Create a table using tabulate.
+        if what == "owned":
+            headers = ["POS", "User", "Assets Owned"]
+        else #what = default: order by value of stock worth
+            headers = ["POS", "User", "Total Value"]
+        table = []
+
+        for n, u in enumerate(sorted_list[:10]):
+            try:
+                member = ctx.message.guild.get_member(u.id).name
+                # Unicode and tables suck
+                member = member.encode("ascii", errors="replace").decode()
+            except AttributeError:
+                # Prevent race condition - member leaving between command invocation and here
+                continue
+            stocks = await ctx.bot.database.get_user_stocks(u)
```

Comment:
The first call to get the stocks was to help sort the list, the second call was to get and use the stocks... Maybe I could append a column onto the list...